packages/kbn-user-profile-components/src/user_profiles_selectable.tsx

```diff
@@ -0,0 +1,264 @@
+  selectedOptions,
+  defaultOptions,
+  options,
+  onChange,
+  onSearchChange,
+  isLoading = false,
+  singleSelection = false,
+  height,
+}) => {
+  const [displayedOptions, setDisplayedOptions] = useState<SelectableOption[]>([]);
+
+  // Resets all displayed options
+  const resetDisplayedOptions = () => {
+    if (options) {
+      setDisplayedOptions(options.map(toSelectableOption));
+      return;
+    }
+
+    setDisplayedOptions([]);
+    updateDisplayedOptions();
+  };
+
+  const ensureSeparator = (values: SelectableOption[]) => {
+    let index = values.findIndex((option) => option.isGroupLabel);
+    if (index === -1) {
+      const length = values.push({
+        label: i18n.translate('userProfileComponents.userProfilesSelectable.suggestedLabel', {
+          defaultMessage: 'Suggested',
+        }),
+        isGroupLabel: true,
+      } as SelectableOption);
+      index = length - 1;
+    }
+    return index;
+  };
+
+  // Updates displayed options without removing or resorting exiting options
+  const updateDisplayedOptions = () => {
+    if (options) {
+      return;
+    }
+
+    setDisplayedOptions((values) => {
+      // Copy all displayed options
+      const nextOptions: SelectableOption[] = [...values];
+
+      // Get any newly added selected options
+      const selectedOptionsToAdd: SelectableOption[] = selectedOptions
+        ? selectedOptions
+            .filter((profile) => !nextOptions.find((option) => option.key === profile.uid))
+            .map(toSelectableOption)
+        : [];
+
+      // Get any newly added default options
+      const defaultOptionsToAdd: SelectableOption[] = defaultOptions
+        ? defaultOptions
+            .filter(
+              (profile) =>
+                !nextOptions.find((option) => option.key === profile.uid) &&
+                !selectedOptionsToAdd.find((option) => option.key === profile.uid)
+            )
+            .map(toSelectableOption)
+        : [];
+
+      // Merge in any new options and add group separator if necessary
+      if (defaultOptionsToAdd.length) {
+        const separatorIndex = ensureSeparator(nextOptions);
+        nextOptions.splice(separatorIndex, 0, ...selectedOptionsToAdd);
+        nextOptions.push(...defaultOptionsToAdd);
+      } else {
+        nextOptions.push(...selectedOptionsToAdd);
+      }
+
+      return nextOptions;
+    });
+  };
+
+  // Marks displayed options as checked or unchecked depending on `props.selectedOptions`
+  const updateCheckedStatus = () => {
+    setDisplayedOptions((values) =>
+      values.map((option) => {
+        if (selectedOptions) {
+          const match = selectedOptions.find((p) => p.uid === option.key);
+          return { ...option, checked: match ? 'on' : undefined };
+        }
+        return { ...option, checked: undefined };
+      })
+    );
+  };
+
+  useEffect(resetDisplayedOptions, [options]); // eslint-disable-line react-hooks/exhaustive-deps
```

Comment:
Hey! To my understanding, the following three `useEffect` change the `displayedOptions` based on prop changes. If that is the case, we should avoid them (to avoid rerenders) and call the functions directly in the component.

Reply:
This component is a bit of a strange one since it's controlled (we want to render whatever options have been passed in as selected or suggested) but also uncontrolled (options should not change order while displayed). This is so that e.g. a user doesn't jump up into the "selected" group when being toggled so the behaviour here is correct.

Reply:
I see what you mean. Maybe I am wrong, but do you think the following would achieve the same effect (maybe with some extra checking inside the functions)?

```
const [displayedOptions, setDisplayedOptions] = useState<SelectableOption[]>([]);

// rest of the code

resetDisplayedOptions()
updateDisplayedOptions()
updateCheckedStatus()

const selectedCount = selectedOptions ? selectedOptions.length : 0;

return (...)
```

Reply:
That would not work since it would reset the list of displayed options on every render. We only want to do a hard reset when `options` prop changes (so that we show search results instead of default options). When only `selectedOptions` prop changes we don't want to do a hard reset. Instead we should leave the list as is and only update the `checked` status. Otherwise we would get the jumping options issue I talked about above. 

What's your concern with the `useEffect` call?

Reply:
My main concern with the `useEffect` is that they run after the changes have been committed by React to the DOM. That means that the `useEffect`s will cause another rerender cycle because they update the state of the component. To my understanding so far, we derive states solely from the props. I think that the state can be derived without the use of the `useEffect`s. I am not saying is wrong or we should change it. I just wanted to talk about it 🙂.

Reply:
Definitely, thanks for raising your concerns! 

> To my understanding so far, we derive states solely from the props

If that was the case you would be definitely be correct and we could simply derive displayed options from props without `useEffect`. However, this is a stateful component since we need to keep track of our own internal list to implement the desired UX:

```
const [displayedOptions, setDisplayedOptions] = useState<SelectableOption[]>([]);
```

I know it's not particular pretty but I'm not sure how we can implement the desired behaviour without `useEffect` since we need to recalculate our internal state based on prop changes.

Reply:
Thank you for your patience! You are right. I think the only way to do it is to keep track of [previous values](https://reactjs.org/docs/hooks-faq.html#how-do-i-implement-getderivedstatefromprops) and compare them with the current values before updating the state. For example:

```
// code is simplified for brevity
const [displayedOptions, setDisplayedOptions] = useState<SelectableOption[]>([]);
const [prevProps, setPrevProps] = useState({ options, defaultOptions, selectedOptions }) // you can also use useRef

if (!deepEqual(prevProps.options, options)) {
    resetDisplayedOptions()
    setPrevProps({ options, defaultOptions, selectedOptions })
}

if (!deepEqual(prevProps.options, options) || !deepEqual(prevProps.selectedOptions, selectedOptions)) {
    updateDisplayedOptions()
    setPrevProps({ options, defaultOptions, selectedOptions })
}

// same for updateCheckedStatus
```

Though the code may not be that readable 🙂.

Reply:
My concern with that approach is that we'd be running deep equality checks on every render which isn't free either. I think we should be careful trying to blindly optimise React without benchmarking as it's easy to make things worse. 

I'm currently not convinced that would be worth the effort though. Are you aware of any use cases where there will be many of these components rendered on the page at the same time? I've only seen the use cases inside a popover.

Reply:
Ok @thomheymann! I totally agree with not optimizing anything without benchmarking. For cases, the components will be used one time on each page.